pkg/relayer/session/persistence.go

```diff
@@ -0,0 +1,200 @@
+package session
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/pokt-network/poktroll/pkg/relayer"
+	prooftypes "github.com/pokt-network/poktroll/x/proof/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+// populateSessionTreeMap reconstructs the in-memory session tree map from previously persisted sessions.
+// This method:
+// 1. Retrieves all persisted session metadata from storage
+// 2. Validates each session against current blockchain height to determine its lifecycle state
+// 3. Deletes expired sessions that are past their proof submission window
+// 4. Deletes sessions past their claim window that don't have an associated onchain claim
+// 5. Retains and loads valid sessions into memory for continued processing
+func (rs *relayerSessionsManager) populateSessionTreeMap(ctx context.Context, height int64) error {
+	logger := rs.logger.With("method", "populateSessionTreeMap", "height", height)
+
+	sharedParams, err := rs.sharedQueryClient.GetParams(ctx)
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to get shared params")
+		return err
+	}
+
+	// Retrieve all persisted session metadata for evaluation of their current state
+	_, persistedSessions, err := rs.sessionsMetadataStore.GetAll([]byte{}, false)
+	if err != nil {
+		logger.Error().Err(err).Msg("failed to get persisted sessions")
+		return err
+	}
+
+	for _, persistedSession := range persistedSessions {
+		// Unmarshal the persisted session metadata for processing
+		persistedSMT := &prooftypes.PersistedSMT{}
+		if err := persistedSMT.Unmarshal(persistedSession); err != nil {
+			logger.Error().Err(err).Msg("failed to unmarshal persisted session metadata, skipping")
+			continue
+		}
+
+		sessionEndHeight := persistedSMT.SessionHeader.SessionEndBlockHeight
+		sessionId := persistedSMT.SessionHeader.SessionId
+		supplierOperatorAddress := persistedSMT.SupplierOperatorAddress
+
+		// Create a contextual logger for this specific session
+		sessionLogger := logger.With(
+			"session_id", sessionId,
+			"supplier_operator_address", supplierOperatorAddress,
+			"session_end_height", sessionEndHeight,
+		)
+
+		// Session lifecycle states and corresponding actions:
+		// 1. Active and accepting relays: Retain session for processing new relays
+		// 2. Within claim window: Retain session as it's in the process of creating a claim
+		// 3. Past claim window with onchain claim: Retain for proof submission
+		// 4. Past claim window without onchain claim: Delete as it's too late for settlement
+		// 5. Past proof window: Delete as the session is completely expired
```

Comment:
Perhaps there's an opportunity here to log whether the (and which) supplier should be expecting to be slashed in relevant cases (claimed + proof window closed + proof required). Even if it's just a TODO for now, wdyt?